ls1intum/Artemis
# `Exam mode`: Speedup unlock operation for programming exercises

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Unlocking many repositories could take some time. This PR aims to speedup the process.

### Description
The performance increments come from:
- Reducing the needed calls to the external system from 2 to 1 (we only configure the branch settings once, that's not needed every time).
- Removing an unnecessary database call that was executed for every participation.

This cuts the time needed for the unlock operation in half!

The other changes in this PR are refactorings:
- Since the unlock operations don't need the programming exercise, we removed the parameter. This should avoid unnecessary database operations in the future.
- There were two duplicated enums containing repository permissions. We now removed one of them.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Exam with many participations

1. Use prepare exercises to create the repositories
2. Verify that branch protection rules are in place. Verify that the students cannot push to the repository.
3. Use the "unlock repositories" button or wait until the planed start of the exercise
4. Verify that all participations got successfully unlocked.

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests
